xgcm/autogenerate.py

```diff
@@ -0,0 +1,254 @@
+        e.g. (dims={'X':'lon','Y':'lon'}). These have to be names of
+        strictly dimensions of ds.
+
+        to - {'left','right'} Direction of shift, similar as the logic in
+        xgcm.Axis, needed to apply shift correctly
+
+        x_wrap - float: value of discontinuity in x direction (defaults to
+        360 for global longitude)
+
+        y_wrap - float: as x_wrap but for y direction (defaults to
+        180 for global longitude)
+
+        z_pad - float: boundary padding value for the depth interpolation. If
+        not specified it will default to the min() or max() of the appropriate
+        field, depending on 'to'
+
+        dim_switch - Bool: switch between input for dimesions (need to
+        load class from xgcm.autogenerate) and coordinates (uses existing
+        xgc.grid logic)
+        """
+
+        #TODO: I can definitely infer the dims from the comodo stuff, but for
+        #now leave it explicit
+
+        for ki in axes.keys():
+            if ki not in dims.keys():
+                raise RuntimeError('dimension '+ki+' was not specified in axes \
+                                    input')
+            old_name = axes[ki]
+            dim = dims[ki]
+            new_name = old_name+'_inferred'
+
+            if ki == 'X':
+                periodic = True
+                boundary = None
+                fill_value = 0.0
+                wrap = x_wrap
+            elif ki == 'Y':
+                periodic = True
+                boundary = None
+                fill_value = 0.0
+                wrap = y_wrap
+            elif ki == 'Z':
+                periodic = False
+                boundary = 'fill'
+                # For depth decide between surface or bottom fill
+                if z_pad is None:
+                    # Infer limits from depth data
+                    z_min = ds[old_name].min().data
+                    z_max = ds[old_name].max().data
+                    # The value below is an attempt on extrapolating the values
+                    # linearly. The difference between values at the top and
+                    # bottom is used to pad the depth array at the appropriate
+                    # boundary. For multidimensional arrays the min difference
+                    # is chosen. This could lead to undesired results if the
+                    # depth spacing is spatially irregular (though thats a rare
+                    # case IMO)
+                    top_diff = ds[old_name].diff(dim).isel(**{dim: 0}).min()
+                    bot_diff = ds[old_name].diff(dim).isel(**{dim: -1}).min()
+                    top = z_min.data - top_diff.data
+                    bot = z_max.data + bot_diff.data
+                    if to is 'left':
+                        fill_value = top
+                    elif to is 'right':
+                        fill_value = bot
+                else:
+                    fill_value = z_pad
```

Comment:
I'm not sure if the design of this function is optimal. Could we not just have a generic `generate_axis` function that has sufficient flexibility via kwargs to apply to any scenario? Then we would loop over axes in `autogenerate_ds` and call this `generate_axis` function on each.